axum/src/ext_traits/request.rs

```diff
@@ -0,0 +1,200 @@
+pub trait RequestExt<B>: sealed::Sealed<B> + Sized {
+    /// Apply an extractor to this `Request`.
+    ///
+    /// This is just a convenience for `E::from_request(req, &())`.
+    ///
+    /// Note this consumes the request. Use [`RequestExt::extract_parts`] if you're not extracting
+    /// the body and don't want to consume the request.
+    fn extract<E, M>(self) -> BoxFuture<'static, Result<E, E::Rejection>>
+    where
+        E: FromRequest<(), B, M> + 'static,
+        M: 'static;
+
+    /// Apply an extractor that requires some state to this `Request`.
+    ///
+    /// This is just a convenience for `E::from_request(req, state)`.
+    ///
+    /// Note this consumes the request. Use [`RequestExt::extract_parts_with_state`] if you're not
+    /// extracting the body and don't want to consume the request.
+    fn extract_with_state<E, S, M>(self, state: &S) -> BoxFuture<'_, Result<E, E::Rejection>>
+    where
+        E: FromRequest<S, B, M> + 'static,
+        S: Send + Sync;
+
+    /// Apply a parts extractor to this `Request`.
+    ///
+    /// This is just a convenience for `E::from_request_parts(parts, state)`.
+    fn extract_parts<E>(&mut self) -> BoxFuture<'_, Result<E, E::Rejection>>
+    where
+        E: FromRequestParts<()> + 'static;
+
+    /// Apply a parts extractor that requires some state to this `Request`.
+    ///
+    /// This is just a convenience for `E::from_request_parts(parts, state)`.
+    fn extract_parts_with_state<'a, E, S>(
+        &'a mut self,
+        state: &'a S,
+    ) -> BoxFuture<'a, Result<E, E::Rejection>>
+    where
+        E: FromRequestParts<S> + 'static,
+        S: Send + Sync;
+}
+
+impl<B> RequestExt<B> for Request<B>
+where
+    B: Send + 'static,
+{
+    fn extract<E, M>(self) -> BoxFuture<'static, Result<E, E::Rejection>>
+    where
+        E: FromRequest<(), B, M> + 'static,
+        M: 'static,
+    {
+        self.extract_with_state(&())
+    }
+
+    fn extract_with_state<E, S, M>(self, state: &S) -> BoxFuture<'_, Result<E, E::Rejection>>
+    where
+        E: FromRequest<S, B, M> + 'static,
+        S: Send + Sync,
+    {
+        E::from_request(self, state)
+    }
+
+    fn extract_parts<E>(&mut self) -> BoxFuture<'_, Result<E, E::Rejection>>
+    where
+        E: FromRequestParts<()> + 'static,
+    {
+        self.extract_parts_with_state(&())
+    }
+
+    fn extract_parts_with_state<'a, E, S>(
+        &'a mut self,
+        state: &'a S,
+    ) -> BoxFuture<'a, Result<E, E::Rejection>>
+    where
+        E: FromRequestParts<S> + 'static,
+        S: Send + Sync,
+    {
+        let mut req = Request::new(());
+        *req.version_mut() = self.version();
+        *req.method_mut() = self.method().clone();
+        *req.uri_mut() = self.uri().clone();
+        *req.headers_mut() = std::mem::take(self.headers_mut());
+        *req.extensions_mut() = std::mem::take(self.extensions_mut());
+        let (mut parts, _) = req.into_parts();
```

Comment:
Hm, this is kinda unfortunate. As an alternative, we could require `B: Default` and use `mem::take`, wdyt about that? Pretty much everything in `http-body` including the boxed body types seems to implement `Default` if the inner buffer does, the only exception being `Limited` but I'm sure there it can be added too.

Reply:
Yeah thats probably fine. I'll change it.

Reply:
Hm actually its kinda weird having to write this on your extractor. Feels kinda arbitrary from the user perspective 🤔

```diff
@@ -180,7 +171,7 @@ mod tests {
     impl<S, B> FromRequest<S, B> for WorksForCustomExtractor
     where
         S: Send + Sync,
-        B: Send + 'static,
+        B: Send + Default + 'static,
         String: FromRef<S> + FromRequest<(), B>,
     {
         type Rejection = <String as FromRequest<(), B>>::Rejection;
```

Reply:
Right, I didn't consider that you'd want to use this from functions generic over `B`. Maybe `http` can be updated to provide `.parts(&self) -> &Parts` and `.parts_mut(&mut self) -> &mut Parts` on `Request` and `Response`, the current implementation of the types would trivially allow this.

Reply:
I remember there was some discussion around that a while ago https://github.com/hyperium/http/pull/511.